// myges/dist/commands/login.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const commands_base_1 = require("../commands-base");
const configurator = __importStar(require("../config"));
function register(program) {
    program
        .command('login')
        .description('sign in to an account')
        .action((0, commands_base_1.errorHandler)(action));
}
exports.register = register;
async function action(options) {
    try {
        const token = await configurator.prompt_credentials();
        await configurator.save({
            username: token.username,
            token_type: token.token_type,
            access_token: token.access_token,
            expires: Date.now() + parseInt(token.expires_in, 10) * 1000,
        });
        console.log('Successfully logged in!');
    }
    catch (e) {
        if (options.debug) {
            console.error(e);
        }
        else {
            console.error(e.message);
        }
    }
}
